### PR TITLE
Add jet reconstruction

### PR DESCRIPTION
**Pull request by ruse-traveler**
### Briefly, what does this PR introduce?

This PR introduces basic jet-reconstruction capabilities to EICrecon.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [X ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [X ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

No.

### Does this PR change default behavior?

Yes. Two new output collections will be created by default: a collection of jets made from reconstructed charged particles and their parameters, and a collection of jets made from generator-level particles and their parameters.

### CODE REVIEW

File: src/algorithms/reco/JetReconstruction.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#ifndef EICRECON_JETRECONSTRUCTION_H
+#define EICRECON_JETRECONSTRUCTION_H
+
+// standard c includes
+#include <vector>
+// DD4hep includes
+#include <spdlog/spdlog.h>
+#include <DD4hep/DD4hepUnits.h>
+#include <edm4hep/utils/kinematics.h>
+#include <edm4eic/MutableReconstructedParticle.h>
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.h:13]: The MutableReconstructedParticle.h needs to be included in the .cc file.
```suggestion
#include <edm4eic/ReconstructedParticle.h>
```

File: src/algorithms/reco/JetReconstruction.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      jet_edm.setMomentum(edm4hep::Vector3f(jets[i].px(), jets[i].py(), jets[i].pz()));
+      jet_edm.setEnergy(jets[i].e());
+      jet_edm.setMass(jets[i].m());
+
+      std::vector<PseudoJet> csts = jets[i].constituents();
+      for (unsigned j = 0; j < csts.size(); j++) {
+        const double cst_pt = csts[j].pt();
+        m_log->trace("    constituent {}'s pt: {}", j, cst_pt);
+        // Only consider constituents in the momentum range
+        if (cst_pt > m_minCstPt && cst_pt < m_maxCstPt) {
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:69]: For readability:
```suggestion
        if ((cst_pt > m_minCstPt) && (cst_pt < m_maxCstPt)) {
```

File: src/algorithms/reco/JetReconstruction.h (hunk 1 above)
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.h:13]: Gotcha! So "ReconstructedParticle.h" in the .h file and "MutableReconstructedParticle.h" in the .cc file?

File: src/algorithms/reco/JetReconstruction.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+#ifndef EICRECON_JETRECONSTRUCTION_H
+#define EICRECON_JETRECONSTRUCTION_H
+
+// standard c includes
+#include <vector>
+// DD4hep includes
+#include <spdlog/spdlog.h>
+#include <DD4hep/DD4hepUnits.h>
+#include <edm4hep/utils/kinematics.h>
```
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.h:12]: I've also removed the headers for things we aren't using right now.

File: src/algorithms/reco/JetReconstruction.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    m_log->trace("  Clustering with : {}", jet_def.description());
+
+    for (unsigned i = 0; i < jets.size(); i++) {
+
+      m_log->trace("  jet {}: pt = {}, y = {}, phi = {}", i, jets[i].pt(), jets[i].rap(), jets[i].phi());
+      edm4eic::MutableReconstructedParticle jet_edm;
+      // Type = 0 for jets, Type = 1 for constituents
+      // Use PDG values to match jets and constituents
+      jet_edm.setType(0);
+      jet_edm.setPDG(i);
```
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/algorithms/reco/JetReconstruction.cc:59]: This is a stopgap while PODIO's one-to-many relations aren't working. The rationale is that we would like to store the constituents of the reconstructed jets while preserving the relation between them, and our solution was to store both the jets and the constituents in the same collection. So we use `Type` field to indicate what's what (0 for a reconstructed jet, and 1 for a jet constituent) and the `PDG` field to store the index of the jet (to associate each constituent with its jet). 

(As an aside: we'll be documenting this on the wiki for posterity!)

[the remark this answers is not in the document]
Comment (veprbl) [reply src/algorithms/reco/JetReconstruction.cc:63]: Oh, on the second look, I see the comment on the line before last.

File: src/services/io/podio/JEventProcessorPODIO.cc [hunk 5]
```diff
@@ -109,6 +109,8 @@ JEventProcessorPODIO::JEventProcessorPODIO() {
             "InclusiveKinematicseSigma",
             "InclusiveKinematicsElectron",
             "InclusiveKinematicsTruth",
+            "JetTruth",
+            "JetReco",
```
Comment (veprbl) [inline src/services/io/podio/JEventProcessorPODIO.cc:113]:
```suggestion
            "GeneratedJets",
            "ReconstructedJets",
```

File: src/global/reco/reco.cc [hunk 6]
```diff
@@ -19,6 +19,8 @@
 #include "InclusiveKinematicsDA_factory.h"
 #include "InclusiveKinematicseSigma_factory.h"
 #include "InclusiveKinematicsSigma_factory.h"
+#include "JetTruth_factory.h"
+#include "JetReco_factory.h"
```
Comment (veprbl) [inline src/global/reco/reco.cc:23]:
```suggestion
#include "GeneratedJets_factory.h"
#include "ReconstructedJets_factory.h"
```

Comment (ruse-traveler) [reply src/global/reco/reco.cc]: Done! (Apologies for the delay!)

File: src/services/io/podio/JEventProcessorPODIO.cc (hunk 5 above)
Comment (ruse-traveler) [reply src/services/io/podio/JEventProcessorPODIO.cc]: Done! (Apologies for the delay!)

File: cmake/FindFastJet.cmake [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#  FASTJET_INCLUDE_DIRS (not cached)
+#  FASTJET_LIBRARY
+#  FASTJET_LIBRARIES (not cached)
+#  FASTJET_LIBRARY_DIRS (not cached)
+
+find_path(FASTJET_INCLUDE_DIR fastjet/version.hh
+          HINTS $ENV{FASTJET_ROOT_DIR}/include ${FASTJET_ROOT_DIR}/include)
+
+find_library(FASTJET_LIBRARY NAMES fastjet
+             HINTS $ENV{FASTJET_ROOT_DIR}/lib ${FASTJET_ROOT_DIR}/lib)
```
Comment (veprbl) [inline cmake/FindFastJet.cmake:15]:
```suggestion
          HINTS $ENV{FASTJET_ROOT}/include ${FASTJET_ROOT_DIR}/include)

find_library(FASTJET_LIBRARY NAMES fastjet
             HINTS $ENV{FASTJET_ROOT}/lib ${FASTJET_ROOT_DIR}/lib)
```

File: src/algorithms/reco/JetReconstruction.cc [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+namespace eicrecon {
+
+  void JetReconstruction::init(std::shared_ptr<spdlog::logger> logger) {
+    m_log = logger;
+    m_log->trace("Initialized");
+  }
+
+  std::vector<edm4eic::ReconstructedParticle*> JetReconstruction::execute(
+      std::vector<const edm4hep::LorentzVectorE*> momenta) {
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.cc:27]:
```suggestion
      const std::vector<const edm4hep::LorentzVectorE*> &momenta) {
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    void GeneratedJets_factory::Process(const std::shared_ptr<const JEvent> &event) {
+        auto mc_particles = event->Get<edm4hep::MCParticle>("MCParticles");
+
+        std::vector<const edm4hep::LorentzVectorE*> momenta;
+        for (const auto& p: mc_particles)
+          if (p->getGeneratorStatus() == 1) {
+            const auto& mom = p->getMomentum();
+            const auto& energy = p->getEnergy();
+            momenta.push_back(new edm4hep::LorentzVectorE(mom.x, mom.y, mom.z, energy));
+          }
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:43]:
```suggestion
        for (const auto& p: mc_particles) {
          if (p->getGeneratorStatus() == 1) {
            const auto& mom = p->getMomentum();
            const auto& energy = p->getEnergy();
            momenta.push_back(new edm4hep::LorentzVectorE(mom.x, mom.y, mom.z, energy));
          }
        }
```

File: src/global/reco/GeneratedJets_factory.cc [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+        auto jets = m_jet_algo.execute(momenta);
```
Comment (veprbl) [inline src/global/reco/GeneratedJets_factory.cc:45]: Avoid a memory leak:
```suggestion
        auto jets = m_jet_algo.execute(momenta);
        for (const auto &mom : momenta) {
          delete mom;
        }
```

File: src/global/reco/ReconstructedJets_factory.cc [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+        std::vector<const edm4hep::LorentzVectorE*> momenta;
+        for (const auto& p: rc_particles) {
+          // TODO: Need to exculde the scattered electron
+          const auto& mom = p->getMomentum();
+          const auto& energy = p->getEnergy();
+          momenta.push_back(new edm4hep::LorentzVectorE(mom.x, mom.y, mom.z, energy));
+        }
+
+        auto jets = m_jet_algo.execute(momenta);
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.cc:43]: Avoid memory leak:
```suggestion
        auto jets = m_jet_algo.execute(momenta);
        for (const auto &mom : momenta) {
          delete mom;
        }
```

File: src/global/reco/ReconstructedJets_factory.cc [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    void ReconstructedJets_factory::ChangeRun(const std::shared_ptr<const JEvent> &event) {
+        // Nothing to do here
+    }
+
+    void ReconstructedJets_factory::Process(const std::shared_ptr<const JEvent> &event) {
+        auto rc_particles = event->Get<edm4eic::ReconstructedParticle>("ReconstructedParticles");
+
+        std::vector<const edm4hep::LorentzVectorE*> momenta;
+        for (const auto& p: rc_particles) {
+          // TODO: Need to exculde the scattered electron
```
Comment (veprbl) [inline src/global/reco/ReconstructedJets_factory.cc:37]:
```suggestion
          // TODO: Need to exclude the scattered electron
```

File: src/algorithms/reco/JetReconstruction.h [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+  // jet reco configuration
+  class JetReconstruction {
+
+    public:
+
+      void init(std::shared_ptr<spdlog::logger> logger);
+
+      std::vector<edm4eic::ReconstructedParticle*> execute(
+          std::vector<const edm4hep::LorentzVectorE*> momenta
```
Comment (veprbl) [inline src/algorithms/reco/JetReconstruction.h:31]:
```suggestion
          const std::vector<const edm4hep::LorentzVectorE*> &momenta
```

**Review by veprbl — APPROVED** [review]
LGTM. The only issue is that you seem to leak memory on an intermediate lorentz vector collection that is not taken for ownership by JANA.

File: src/global/reco/GeneratedJets_factory.cc (hunk 10 above)
Comment (veprbl) [reply src/global/reco/GeneratedJets_factory.cc:45]: Or switch to `std::vector<std::unique_ptr<` ...

Comment (veprbl) [reply src/global/reco/GeneratedJets_factory.cc:46]: Or don't use pointers at all and do `momenta.emplace_back(mom.x, mom.y, mom.z, energy)`

**Review by veprbl — APPROVED** [review]

**Pull request merged**